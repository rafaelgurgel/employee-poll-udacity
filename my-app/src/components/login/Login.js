import React, { useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { handleSetAuthedUser } from "../../actions/shared";
import { useNavigate, useLocation } from "react-router-dom";
import "./login.css";

export default function Login() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const location = useLocation(); // Capture current location

  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");

  const authedUser = useSelector((state) => state.authedUser);
  const users = useSelector((state) => state.users);

  // Get the intended path from location.state or default to "/home"
  const redirectTo = location.state?.from || "/home";

  useEffect(() => {
    if (authedUser) {
      navigate(redirectTo, { replace: true }); // Redirect to the intended path
    }
  }, [authedUser, redirectTo, navigate]);

  const handleSubmit = (e) => {
    e.preventDefault();

    const user = users[username];
    if (!user) {
      setError("User does not exist.");
      return;
    }

    if (user.password !== password) {
      setError("Incorrect password.");
      return;
    }

    dispatch(handleSetAuthedUser(username)); // Set authedUser in Redux state
  };

  return (
    <div className="login-container">
      <h2 className="login-title">Employee Polls Login</h2>

      <form className="login-form" onSubmit={handleSubmit}>
        <label htmlFor="username">Username</label>
        <input
          type="text"
          id="username"
          placeholder="Enter username (e.g. sarahedo)"
          value={username}
          onChange={(e) => {
            setUsername(e.target.value);
            setError("");
          }}
        />

        <label htmlFor="password">Password</label>
        <input
          type="password"
          id="password"
          placeholder="Enter password"
          value={password}
          onChange={(e) => {
            setPassword(e.target.value);
            setError("");
          }}
        />

        <button type="submit">Login</button>
      </form>

      {error && <p className="error-message">{error}</p>}
    </div>
  );
}