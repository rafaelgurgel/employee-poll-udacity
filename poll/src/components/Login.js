import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { handleSetAuthedUser } from '../actions/shared'
import { useNavigate } from 'react-router-dom'

export default function Login() {
  const dispatch = useDispatch()
  const navigate = useNavigate()

  // We'll store both username and password in local component state.
  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')
  const [error, setError] = useState('')

  // We have all users in Redux state (shape: {sarahedo: {id, password, ...}, ...})
  const users = useSelector((state) => state.users)

  const handleSubmit = (e) => {
    e.preventDefault()

    // Check if the user by this username exists
    const user = users[username]
    if (!user) {
      setError('User does not exist.')
      return
    }

    // Check if the password matches
    if (user.password !== password) {
      setError('Incorrect password.')
      return
    }

    // If credentials are correct, setAuthedUser to this user
    dispatch(handleSetAuthedUser(username))
    navigate('/home')
  }

  return (
    <div className="login-container" style={{ marginTop: 50 }}>
      <h2>Employee Polls Login</h2>
      <form onSubmit={handleSubmit} style={{ display: 'flex', flexDirection: 'column', width: 300 }}>
        <label htmlFor="username">Username</label>
        <input
          type="text"
          id="username"
          placeholder="Enter username (e.g. sarahedo)"
          value={username}
          onChange={(e) => {
            setUsername(e.target.value)
            setError('')
          }}
        />

        <label htmlFor="password" style={{ marginTop: 10 }}>Password</label>
        <input
          type="password"
          id="password"
          placeholder="Enter password"
          value={password}
          onChange={(e) => {
            setPassword(e.target.value)
            setError('')
          }}
        />

        <button type="submit" style={{ marginTop: 20 }}>
          Login
        </button>
      </form>
      {error && <p style={{ color: 'red' }}>{error}</p>}
    </div>
  )
}
