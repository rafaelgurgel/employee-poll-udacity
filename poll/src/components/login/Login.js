// src/components/Login.js
import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { handleSetAuthedUser } from '../../actions/shared'
import { useNavigate } from 'react-router-dom'
import './login.css'

export default function Login() {
  const dispatch = useDispatch()
  const navigate = useNavigate()

  // We'll store both username and password in local component state.
  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')
  const [error, setError] = useState('')

  // We have all users in Redux state
  const users = useSelector((state) => state.users)

  const handleSubmit = (e) => {
    e.preventDefault()

    // Check if the user by this username exists
    const user = users[username]
    if (!user) {
      setError('User does not exist.')
      return
    }

    // Check if the password matches
    if (user.password !== password) {
      setError('Incorrect password.')
      return
    }

    // If credentials are correct, setAuthedUser
    dispatch(handleSetAuthedUser(username))
    navigate('/home')
  }

  return (
    <div className="login-container">
      <h2 className="login-title">Employee Polls Login</h2>

      <form className="login-form" onSubmit={handleSubmit}>
        <label htmlFor="username">Username</label>
        <input
          type="text"
          id="username"
          placeholder="Enter username (e.g. sarahedo)"
          value={username}
          onChange={(e) => {
            setUsername(e.target.value)
            setError('')
          }}
        />

        <label htmlFor="password">Password</label>
        <input
          type="password"
          id="password"
          placeholder="Enter password"
          value={password}
          onChange={(e) => {
            setPassword(e.target.value)
            setError('')
          }}
        />

        <button type="submit">
          Login
        </button>
      </form>

      {error && <p className="error-message">{error}</p>}
    </div>
  )
}
