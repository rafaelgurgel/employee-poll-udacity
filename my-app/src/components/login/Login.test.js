// src/components/login/Login.test.js
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { BrowserRouter } from 'react-router-dom'
import '@testing-library/jest-dom'
import Login from './Login'
import {
  SET_AUTHED_USER,
  LOGOUT_AUTHED_USER,
} from '../../actions/authedUser'

function testReducer(
  state = {
    users: {
      sarahedo: { password: 'abc', name: 'Sarah Edo' },
      johndoe: { password: 'xyz', name: 'John Doe' },
    },
    authedUser: null,
  },
  action
) {
  switch (action.type) {
    case SET_AUTHED_USER:
      return {
        ...state,
        authedUser: action.id,
      }
    case LOGOUT_AUTHED_USER:
      return {
        ...state,
        authedUser: null,
      }
    default:
      return state
  }
}

beforeAll(() => {
  jest.spyOn(console, 'warn').mockImplementation((message) => {
    if (!message.includes('React.startTransition')) {
      console.warn(message)
    }
  })
})

afterAll(() => {
  jest.restoreAllMocks()
})

describe('Login Component', () => {
  it('matches snapshot', () => {
    const store = createStore(testReducer)
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('updates username input on change', () => {
    const store = createStore(testReducer)
    const { getByLabelText } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    )

    const usernameInput = getByLabelText(/username/i)
    fireEvent.change(usernameInput, { target: { value: 'sarahedo' } })
    expect(usernameInput.value).toBe('sarahedo')
  })

  it('shows error if user does not exist', () => {
    const customStore = createStore(() => ({ users: {}, authedUser: null }))

    const { getByLabelText, getByText, getByRole } = render(
      <Provider store={customStore}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    )

    const usernameInput = getByLabelText(/username/i)
    fireEvent.change(usernameInput, { target: { value: 'someunknownuser' } })

    const loginButton = getByRole('button', { name: /login/i })
    fireEvent.click(loginButton)

    expect(getByText(/User does not exist./i)).toBeInTheDocument()
  })

  it('redirects to /home if a user is already logged in', () => {
    const preloadedState = {
      users: {
        sarahedo: { password: 'abc', name: 'Sarah Edo' },
      },
      authedUser: 'sarahedo',
    };
    const store = createStore(testReducer, preloadedState);

    const { queryByText } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    expect(queryByText(/Employee Polls Login/i)).not.toBeInTheDocument();
  });

 
  it('removes the user (authedUser) from Redux after logout', () => {
    const preloadedState = {
      users: {
        sarahedo: { password: 'abc' },
      },
      authedUser: 'sarahedo',
    }
    const store = createStore(testReducer, preloadedState)

    store.dispatch({ type: LOGOUT_AUTHED_USER })

    const state = store.getState()
    expect(state.authedUser).toBeNull()
  })
})
