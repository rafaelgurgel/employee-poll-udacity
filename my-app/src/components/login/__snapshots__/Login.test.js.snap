// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <h2
      class="login-title"
    >
      Employee Polls Login
    </h2>
    <form
      class="login-form"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        id="username"
        placeholder="Enter username (e.g. sarahedo)"
        type="text"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        placeholder="Enter password"
        type="password"
        value=""
      />
      <button
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
